apf/apf.cc

```diff
@@ -482,9 +482,37 @@ void getVectorShapeValues(Element* e, Vector3 const& local,
   }
   else
   {
-    // TODO when ref dim != mesh space dim. Pseudo-inverse needed.
-    PCU_ALWAYS_ASSERT_VERBOSE(false,
-    	"not yet implemented for 3D surface meshes (i.e., manifolds)!");
+    // TODO clean up
+    apf::Matrix3x3 J;
+    apf::getJacobian(e->getParent(), local, J);
+
+    apf::Matrix3x3 JT = apf::transpose(J);
+    apf::Matrix3x3 JJT = J * JT;
+    // take inverse of JTJ and transpose it
+    apf::Matrix<2,2> jjt;
+    for (int i = 0; i < 2; i++)
+      for (int j = 0; j < 2; j++)
+        jjt[i][j] = JJT[i][j];
+    apf::Matrix<2,2> JJTinv = apf::invert(jjt);
+    apf::Matrix<2,2> JJTinvT = apf::transpose(JJTinv);
+
+    apf::Matrix<2,3> JJTinvTJ; // JJTinvT * J
+    for( int i = 0; i < 2; i++ ) {
+      for ( int j = 0; j < 3; j++ ) {
+        JJTinvTJ[i][j] = 0.;
+        for ( int k = 0; k < 2; k++ )
+          JJTinvTJ[i][j] += JJTinvT[i][k] * J[k][j];
+      }
+    }
+
+    // u(x_hat) * J(x_hat)^{-1}
+    for( size_t i = 0; i < values.size(); i++ ) {
+      for ( int j = 0; j < 3; j++ ) {
+        values[i][j] = 0.;
+        for ( int k = 0; k < 2; k++ )
+          values[i][j] += vvals[i][k] * JJTinvTJ[k][j];
+      }
+    }
```

Comment:
@Samiullah-Malik 

Pumi already has a routine to compute pseudo inverse of non-square Jacobian (https://www.scorec.rpi.edu/pumi/doxygen/namespaceapf.html#a038f86d6baa21578815f05d0ab592330). Please look at the code there and if what is done here is the same use it instead of the implementation here (Just keep in mind that in pumi Jacobian is actually the transpose of the Jacobian).

Reply:
Yes, I wanted a psuedo inverse here. So, I have used PUMI's routine to do it now. I also checked that the new values agree with my previous values.